hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/pipeline/PipelineManagerV2Impl.java

```diff
@@ -327,6 +349,9 @@ public void openPipeline(PipelineID pipelineId) throws IOException {
    * @throws IOException
    */
   protected void removePipeline(Pipeline pipeline) throws IOException {
+    if (!scmhaManager.isLeader()) {
```

Comment:
Would it be cleaner by using `InvocationHander` to inject these `isLeader()` check ?

```
StateManager pipelineStateManager = (StateManager) Proxy.newProxyInstance(
    null,
    new Class<?>[] {StateManager.class}, 
    (proxy, method, args) -> {
      if (method.getName().equals("xxx")) {
        // do is leader check;
        // throw exception if needed;
      }
      return method.invoke(args);
    });
```

You may use `hashmap` or `annotation` to speed up the method name check.
Just a recommendation.

Reply:
If we want to refactor the code, I would prefer annotation just like @Replicate in which we would create something like CheckLeaderInvocationHandler extends InvocationHandler and do isLeader check there.
It would be convenient in interface. However, We also use isLeader check in some local methods instead of override methods. Let's keep it this fashion for now and we see if we can polish this.